### PR TITLE
fix: redirects optimizations

### PR DESCRIPTION
**Pull request by sbellone**
## Fix: Immediate but slow client redirect

We had an issue with some client redirects. For example, [this page](https://www.algolia.com/blog/hacker-news-search-algolia/) contains the following redirect:
`<meta http-equiv="refresh" content="0; URL='/blog/customers/hacker-news-search-algolia/'" />`

Event if the number of seconds is set to `0`, as the page is heavy the redirection is not complete with the `goto` return, and it fails when we try to do the [first `evaluate`](https://github.com/algolia/renderscript/blob/master/src/lib/tasks/Render.ts#L33):
```
Error: Execution context was destroyed, most likely because of a navigation.
```

To fix that, I only had to move the wait for `minWait` immediately after the `goto`, before doing anything else.

**How to test:**
- Manually change the minWait
- http://localhost:3000/render?url=https%3A%2F%2Fwww.algolia.com%2Fblog%2Fhacker-news-search-algolia%2F&ua=Algolia+Crawler%2F1.0.0

## Optimization: server redirect or immediate client redirect

If the `page.url()` has changed after the `goto`, we can return immediately the `resolvedUrl` without trying to render the body.

**How to test:**
- http://localhost:3000/render?url=http%3A%2F%2Flocalhost%3A3000%2Ftest-website%2Fmeta-refresh.html&ua=Algolia+Crawler%2F1.0.0
- http://localhost:3000/render?url=http%3A%2F%2Flocalhost%3A3000%2F301&ua=Algolia+Crawler%2F1.0.0

## Feat: handle async client redirections

The last thing are clients redirections that are triggered after a few seconds.
Currently, if now `minWait` is set, we'll render the HTML as is:
`<meta http-equiv="refresh" content="5; URL='/test-website/basic.html'" />`

I've now added a check on the presence of this meta tag, and immediately return the redirection if we found one.

**How to test:**
- http://localhost:3000/render?url=http%3A%2F%2Flocalhost%3A3000%2Ftest-website%2Fmeta-refresh-5.html&ua=Algolia+Crawler%2F1.0.0

### CODE REVIEW

File: src/lib/tasks/Render.ts [hunk 1]
```diff
@@ -10,6 +10,7 @@ export class RenderTask extends Task<RenderTaskParams> {
   async process(): Promise<void> {
     const { url, waitTime } = this.params;
     const { page } = this.page;
+    const baseHref = `${url.protocol}//${url.host}`;
```
Comment (bodinsamuel) [inline src/lib/tasks/Render.ts:13]:
```suggestion
    const baseHref = url.origin;
```

File: src/lib/tasks/Render.ts [hunk 2]
```diff
@@ -27,17 +28,57 @@ export class RenderTask extends Task<RenderTaskParams> {
...
+        headers,
+        resolvedUrl: page!.url(),
+      };
+      return;
+    }
+
+    try {
+      const metaRefreshElement = await page!.$('meta[http-equiv="refresh"]');
+      if (metaRefreshElement) {
+        console.log(await metaRefreshElement.jsonValue());
```
Comment (bodinsamuel) [inline src/lib/tasks/Render.ts:52]:
```suggestion
```

File: src/lib/tasks/Render.ts [hunk 3]
```diff
@@ -27,17 +28,57 @@ export class RenderTask extends Task<RenderTaskParams> {
...
+        const metaRefreshContent = await metaRefreshElement.getProperty(
+          'content'
+        );
+        const refreshContent = await metaRefreshContent?.jsonValue<string>();
+        const match = refreshContent?.match(/\d+;\s(?:url|URL)=(.*)/);
```
Comment (bodinsamuel) [inline src/lib/tasks/Render.ts:57]: regex could be a constant +
```suggestion
        const match = refreshContent.match(/\d+;\s(?:url|URL)=(.*)/);
```

File: src/lib/tasks/Render.ts [hunk 4]
```diff
@@ -27,17 +28,57 @@ export class RenderTask extends Task<RenderTaskParams> {
...
+    }
+
+    try {
+      const metaRefreshElement = await page!.$('meta[http-equiv="refresh"]');
+      if (metaRefreshElement) {
+        console.log(await metaRefreshElement.jsonValue());
+        const metaRefreshContent = await metaRefreshElement.getProperty(
+          'content'
+        );
+        const refreshContent = await metaRefreshContent?.jsonValue<string>();
```
Comment (bodinsamuel) [inline src/lib/tasks/Render.ts:56]:
```suggestion
        const refreshContent = await metaRefreshContent.jsonValue<string>();
```

File: src/lib/tasks/Render.ts [hunk 5]
```diff
@@ -27,17 +28,57 @@ export class RenderTask extends Task<RenderTaskParams> {
...
+            resolvedUrl: redirectURL,
+          };
+          return;
+        }
+      }
+    } catch (e) {
+      console.log(
+        'Error while trying to check for meta[http-equive="refresh"]',
+        e
+      );
```
Comment (bodinsamuel) [inline src/lib/tasks/Render.ts:76]: should be reported instead

**Review by bodinsamuel — APPROVED** [review]
LGTM

File: src/lib/tasks/Render.ts (hunk 5 above)
Comment (sbellone) [reply src/lib/tasks/Render.ts:76]: Where and how?

File: src/lib/tasks/Render.ts (hunk 4 above)
Comment (sbellone) [reply src/lib/tasks/Render.ts:56]: `TS2532: Object is possibly 'undefined'`

File: src/lib/tasks/Render.ts (hunk 3 above)
Comment (sbellone) [reply src/lib/tasks/Render.ts:57]: `TS2532: Object is possibly 'undefined'.`

File: src/lib/tasks/Render.ts (hunk 4 above)
Comment (bodinsamuel) [reply src/lib/tasks/Render.ts:55]: ah yes I thought it was metaRefreshElement 👌🏻

File: src/lib/tasks/Render.ts (hunk 5 above)
Comment (bodinsamuel) [reply src/lib/tasks/Render.ts:75]: I thought we had sentry in this, but we don't. 
We should note that in Jira imo

**Review by bodinsamuel — APPROVED** [review]

File: src/lib/tasks/Render.ts (hunk 5 above)
Comment (sbellone) [reply src/lib/tasks/Render.ts:75]: The repo is public :)

Comment (bodinsamuel) [reply src/lib/tasks/Render.ts:75]: and ? env var can be passed at runtime and we already have datadog inside

Comment (sbellone) [reply src/lib/tasks/Render.ts:75]: Ok I didn't know, I liked the idea to not bloat the project too much.

**Pull request merged**